src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs

```diff
@@ -537,18 +539,42 @@ void MoveReturnEndToNextBlock()
                     _readerAwaitable.SetUncompleted();
                 }
 
-                while (returnStart != null && returnStart != returnEnd)
+                success = _operationState.TryEndRead();
+            }
+
+            if (!success)
```

Comment:
```suggestion
            if (success)
```

This might be the cause of the test failures.

Reply:
This too 😅